typescript/token/test/HypERC20.t.sol

```diff
@@ -0,0 +1,366 @@
+    }
+
+    function testRemoteTransfer_invalidAmount() public {
+        vm.expectRevert("ERC20: burn amount exceeds balance");
+        _expectRemoteTransfer(REQUIRED_VALUE, TRANSFER_AMT * 11);
+        assertEq(erc20Token.balanceOf(ALICE), 1000e18);
+    }
+
+    function testRemoteTransfer_withCustomGasConfig() public {
+        _setCustomGasConfig();
+
+        uint256 balanceBefore = erc20Token.balanceOf(ALICE);
+        _expectRemoteTransferAndGas(
+            REQUIRED_VALUE,
+            TRANSFER_AMT,
+            GAS_LIMIT * IGP_GAS_PRICE
+        );
+        assertEq(erc20Token.balanceOf(ALICE), balanceBefore - TRANSFER_AMT);
+    }
+}
+
+contract HypERC20CollateralTest is HypTokenTest {
+    using TypeCasts for address;
+    HypERC20Collateral internal erc20Collateral;
+
+    function setUp() public override {
+        super.setUp();
+
+        localToken = new HypERC20Collateral(address(primaryToken));
+        erc20Collateral = HypERC20Collateral(address(localToken));
+
+        HypERC20Collateral(address(localToken)).initialize(
+            address(localMailbox)
+        );
+
+        erc20Collateral.enrollRemoteRouter(
+            DESTINATION,
+            address(remoteToken).addressToBytes32()
+        );
+
+        primaryToken.transfer(address(localToken), 1000e18);
+        primaryToken.transfer(ALICE, 1000e18);
+
+        _enrollRemoteTokenRouter();
+    }
+
+    function testInitialize_revert_ifAlreadyInitialized() public {
+        vm.expectRevert("Initializable: contract is already initialized");
+        erc20Collateral.initialize(ALICE);
+    }
+
+    function testRemoteTransfer() public {
+        uint256 balanceBefore = localToken.balanceOf(ALICE);
+
+        vm.prank(ALICE);
+        primaryToken.approve(address(localToken), TRANSFER_AMT);
+        _expectRemoteTransferWithEmit(REQUIRED_VALUE, TRANSFER_AMT, 0);
+        assertEq(localToken.balanceOf(ALICE), balanceBefore - TRANSFER_AMT);
+    }
+
+    function testRemoteTransfer_invalidAllowance() public {
+        vm.expectRevert("ERC20: insufficient allowance");
+        _expectRemoteTransfer(REQUIRED_VALUE, TRANSFER_AMT);
+        assertEq(localToken.balanceOf(ALICE), 1000e18);
+    }
+
+    function testRemoteTransfer_withCustomGasConfig() public {
+        _setCustomGasConfig();
+
+        uint256 balanceBefore = localToken.balanceOf(ALICE);
+
+        vm.prank(ALICE);
+        primaryToken.approve(address(localToken), TRANSFER_AMT);
+        _expectRemoteTransferAndGas(
+            REQUIRED_VALUE,
+            TRANSFER_AMT,
+            GAS_LIMIT * IGP_GAS_PRICE
+        );
+        assertEq(localToken.balanceOf(ALICE), balanceBefore - TRANSFER_AMT);
+    }
+}
+
+contract HypNativeTest is HypTokenTest {
+    using TypeCasts for address;
+    HypNative internal nativeToken;
+
+    function setUp() public override {
+        super.setUp();
+
+        localToken = new HypNative();
+        nativeToken = HypNative(address(localToken));
```

Comment:
ditto
why alias?

Reply:
nativeToken instead of HypNative(address(localToken))